src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/blob.py

```diff
@@ -0,0 +1,144 @@
+    # 2. Support connection string for source
+    # 3. stop using 'baseblobservice.exists' function. it doesn't provide performance gain
+    #    since it invoked the get_blob_properties any way.
+
+    # Question:
+    # 1. Performance of creating a source blob service
+
+    if pattern is None and not recursive:
+        # TODO: use a validator?
+        raise ValueError('Neither --recursive or --pattern is set.')
+
+    if exclude_old and exclude_new:
+        raise ValueError('Both --exclude-new and --exclude-old are set.')
+
+    src_client = BlockBlobService(account_name=source_account, sas_token=source_sas)
+
+    def _get_blob_name(source_blob):
+        name = source_blob.name
+        if source_blob.snapshot is not None:
+            # the snapshot time string has seven digital in the microseconds, which make strptime
+            # nearly unusable. therefore characters after dot are thrown away
+            time_string = source_blob.snapshot[:source_blob.snapshot.rfind('.')]
+            snapshot_time = datetime.strptime(time_string, '%Y-%m-%dT%H:%M:%S')
+
+            # insert the date time string before the file extension
+            dot = name.rfind('.')
+            dot = len(name) if dot == -1 else dot
+            name = '{0}({1}){2}'.format(name[0:dot],
+                                        snapshot_time.strftime('%Y-%m-%d %H%M%S'),
+                                        name[dot:])
+        return name
+
+    def _get_blob_url(source_blob):
+        src_url = src_client.make_blob_url(source_container, source_blob.name, sas_token=source_sas)
+
+        if source_blob.snapshot is not None:
+            # this is a blob snapshot
+            if '?' in src_url:
+                src_url += '&snapshot=' + str(source_blob.snapshot)
+            else:
+                src_url += '?snapshot=' + str(source_blob.snapshot)
+
+        return src_url
+
+    def _copy_single_blob(source_blob):
+        kwargs = {
+            "container_name": destination_container,
+            "blob_name": _get_blob_name(source_blob),
+            "copy_source": _get_blob_url(source_blob)
+        }
+
+        if (exclude_new or exclude_old) and client.exists(destination_container, source_blob.name):
+            if exclude_old:
+                destination_blob = client.get_blob_properties(destination_container,
+                                                              source_blob.name)
+                kwargs["source_if_modified_since"] = destination_blob.properties.last_modified
+            if exclude_new:
+                kwargs["destination_if_modified_since"] = source_blob.properties.last_modified
+
+        return client.copy_blob(**kwargs)
+
+    if recursive:
+        source_blobs = src_client.list_blobs(
+            source_container,
+            prefix=pattern,
+            include=Include(snapshots=True) if snapshots else None)
+    elif src_client.exists(source_container, pattern):
```

Comment:
The name is misleading. It is not pattern but just a prefix. I'll change it in next commit. 

The given pattern (prefix) doesn't support wildcard, therefore the command won't be able to infer if this is a recursive copy. The conclusion we have today (with @JasonRShaver) is that the recursive copy across container could be dangerous, we need user to explicitly ask the command to do so.